alibi_detect/cd/keops/learned_kernel.py

```diff
@@ -0,0 +1,341 @@
+        """
+        preprocess_batch_fn = self.train_kwargs['preprocess_fn']
+        if isinstance(preprocess_batch_fn, Callable):  # type: ignore[arg-type]
+            x_all = preprocess_batch_fn(x_all)  # type: ignore[operator]
+        if self.has_proj:
+            x_all_proj = predict_batch(x_all, self.kernel.proj, device=self.device, batch_size=self.batch_size_predict,
+                                       dtype=x_all.dtype if isinstance(x_all, torch.Tensor) else torch.float32)
+        else:
+            x_all_proj = x_all
+
+        x, x2, y, y2 = None, None, None, None
+        k_xx, k_yy, k_xy = [], [], []
+        for batch in range(self.n_batches):
+            i, j = batch * self.batch_size_perms, (batch + 1) * self.batch_size_perms
+            # Stack a batch of permuted reference and test tensors and their projections
+            x_proj = torch.cat([x_all_proj[perm[:m]][None, :, :] for perm in perms[i:j]], 0)
+            y_proj = torch.cat([x_all_proj[perm[m:]][None, :, :] for perm in perms[i:j]], 0)
+            if self.has_kernel_b:
+                x = torch.cat([x_all[perm[:m]][None, :, :] for perm in perms[i:j]], 0)
+                y = torch.cat([x_all[perm[m:]][None, :, :] for perm in perms[i:j]], 0)
+            if batch == 0:
+                x_proj = torch.cat([x_all_proj[None, :m, :], x_proj], 0)
+                y_proj = torch.cat([x_all_proj[None, m:, :], y_proj], 0)
+                if self.has_kernel_b:
+                    x = torch.cat([x_all[None, :m, :], x], 0)  # type: ignore[call-overload]
+                    y = torch.cat([x_all[None, m:, :], y], 0)  # type: ignore[call-overload]
+            x_proj, y_proj = x_proj.to(self.device), y_proj.to(self.device)
+            if self.has_kernel_b:
+                x, y = x.to(self.device), y.to(self.device)
+
+            # Batch-wise kernel matrix computation over the permutations
+            with torch.no_grad():
+                x2_proj, x_proj = LazyTensor(x_proj[:, None, :, :]), LazyTensor(x_proj[:, :, None, :])
+                y2_proj, y_proj = LazyTensor(y_proj[:, None, :, :]), LazyTensor(y_proj[:, :, None, :])
+                if self.has_kernel_b:
+                    x2, x = LazyTensor(x[:, None, :, :]), LazyTensor(x[:, :, None, :])
+                    y2, y = LazyTensor(y[:, None, :, :]), LazyTensor(y[:, :, None, :])
+                k_xy.append(self.kernel(x_proj, y2_proj, x, y2).sum(1).sum(1).squeeze(-1))
+                k_xx.append(self.kernel(x_proj, x2_proj, x, x2).sum(1).sum(1).squeeze(-1))
+                k_yy.append(self.kernel(y_proj, y2_proj, y, y2).sum(1).sum(1).squeeze(-1))
+
+        c_xx, c_yy, c_xy = 1 / (m * (m - 1)), 1 / (n * (n - 1)), 2. / (m * n)
+        # Note that the MMD^2 estimates assume that the diagonal of the kernel matrix consists of 1's
+        stats = c_xx * (torch.cat(k_xx) - m) + c_yy * (torch.cat(k_yy) - n) - c_xy * torch.cat(k_xy)
+        return stats[0], stats[1:]
+
+    @staticmethod
+    def trainer(
```

Comment:
Noting that this method is identical to the `torch` one. Since it's a static method anyway, perhaps it makes sense to factor out as a utility function in some `torch`-based module and use for both `torch` and `keops`?

Reply:
Makes sense. Preferred as part of this PR or a quick follow up one (there are a few things as listed above which impact the `torch` version as well and therefore I wanted to do separate in small PRs).

Reply:
Happy to have everything impacting the `torch` backend as a follow-up.